pkg/record/handler/save.go

```diff
@@ -182,75 +182,74 @@ func (h SaveHandler) Handle(req interface{}) (resp interface{}, err error) {
 		ErrMap: map[record.ID]skyerr.Error{},
 	}
 
-	// atomic: Open transaction for all operation
-	// inatomic: Open transaction for extending schema
-	if err = h.TxContext.BeginTx(); err != nil {
-		return
-	}
-
-	// atomic: Close trasaction at the end of function
-	//
-	// inatomic: Transaction should have been closed after extending schema,
-	//           so just handle early return
+	// Open transaction for whole operation if atomic save
 	if payload.Atomic {
+		if err = h.TxContext.BeginTx(); err != nil {
+			return
+		}
+
 		defer func() {
-			if err != nil || len(modifyResp.ErrMap) > 0 {
-				err = h.TxContext.RollbackTx()
-			} else {
-				err = h.TxContext.CommitTx()
-			}
-		}()
-	} else {
-		defer func() {
-			// handle early return
-			if h.TxContext.HasTx() {
-				err = h.TxContext.RollbackTx()
+			if txErr := db.EndTx(h.TxContext, err); txErr != nil {
+				err = txErr
 			}
 		}()
 	}
 
-	// TODO: emit schema updated event
-	_, err = ExtendRecordSchema(h.RecordStore, h.Logger, payload.Records)
-	if err != nil {
-		h.Logger.WithError(err).Errorln("failed to migrate record schema")
-		if myerr, ok := err.(skyerr.Error); ok {
-			err = myerr
+	if err = h.ExtendRecordSchemaWithTx(payload); err != nil {
+		return
+	}
+
+	if err = RecordSaveHandler(&modifyReq, &modifyResp); err != nil {
+		// Override error in atomic save
+		if payload.Atomic && len(modifyResp.ErrMap) > 0 {
+			info := map[string]interface{}{}
+			for recordID, err := range modifyResp.ErrMap {
+				info[recordID.String()] = err
+			}
+
+			err = skyerr.NewErrorWithInfo(skyerr.AtomicOperationFailure,
+				"Atomic Operation rolled back due to one or more errors",
+				info)
 			return
 		}
 
-		err = skyerr.NewError(skyerr.IncompatibleSchema, "failed to migrate record schema")
 		return
 	}
 
-	// When inatomic save: Commit transaction for extending schema
+	results := make([]interface{}, 0, len(payload.RawMaps))
+	h.makeResultsFromIncomingItem(payload.IncomingItems, modifyResp, resultFilter, &results)
+
+	resp = results
+
+	return
+}
+
+// ExtendRecordSchemaWithTx ensure the operation is within a transaction.
+// When the request is inatomic, the operation would be wrapped in a new transaction.
+func (h SaveHandler) ExtendRecordSchemaWithTx(payload SaveRequestPayload) (err error) {
 	if !payload.Atomic {
-		if err = h.TxContext.CommitTx(); err != nil {
+		if err = h.TxContext.BeginTx(); err != nil {
```

Comment:
I am a little confused when the operation is atomic, `BeginTx` in `Handler`, when the operation is inatomic, when have `BeginTx` in `ExtendRecordSchemaWithTx`.

Reply:
`ExtendRecordSchema` executes multiple sql statement, and this function should be atomic no matter if the record save is atomic or not.

At handler level, it should not be aware of whether it need to begin transaction for the underlying store. Either I don't think the store implementation should be aware of that.
 
But I can't find a simple way to do it yet, that's why I put the transaction here 😢.

Reply:
Thanks, I can understand now.

Do you think it would be better if we have an `ensureTx()`? so `ExtendRecordSchemaWithTx` won't be related to `Atomic` idea? 

Just an idea.

Reply:
The `ExtendRecordSchemaWithTx` imo should be related to `atomic`, since at handler level, only when the request is inatomic the function should begin a separate transaction. If the `HasTx` return false in `ExtendRecordSchemaWithTx ` when the request is atomic, it should be considered as programming error.

Reply:
I guess maybe eventually I will just add a `ensureTx()` in every underlying function that execute more than one sql statement 😢.

Reply:
Thanks. I have no further problems. See if @rickmak okay for the changes.